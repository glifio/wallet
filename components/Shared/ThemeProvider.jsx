import React from 'react'
import PropTypes from 'prop-types'
import { ThemeProvider as StyledThemeProvider } from 'styled-components'
import theme from './theme'

/* eslint-disable no-shadow */
const ThemeProvider = ({ children, theme, ...rest }) => {
  return (
    <StyledThemeProvider theme={theme} {...rest}>
      {children}
    </StyledThemeProvider>
  )
}

ThemeProvider.propTypes = {
  theme: PropTypes.object,
  children: PropTypes.node
}

ThemeProvider.defaultProps = {
  theme,
  children: <></>
}

ThemeProvider.displayName = 'ThemeProvider'

export default ThemeProvider
