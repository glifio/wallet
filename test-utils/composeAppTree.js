import React from 'react'
import { FilecoinNumber } from '@openworklabs/filecoin-number'
import { Provider } from 'react-redux'

import { initializeStore } from '../test-utils/index'
import { initialState } from '../store/states'

const defaultPreset = {}

const composeAppTree = (preset = defaultPreset, options = {}) => {
  const state = {}
}
