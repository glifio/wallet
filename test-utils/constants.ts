export const WALLET_ADDRESS = 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi'
