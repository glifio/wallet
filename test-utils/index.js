import { createStore } from 'redux'
import { initialState } from '../store/states'
import reducer from '../store/reducer'

export const flushPromises = () => new Promise(setImmediate)

export function initializeStore(state = initialState) {
  return createStore(reducer, state)
}

export default initializeStore
