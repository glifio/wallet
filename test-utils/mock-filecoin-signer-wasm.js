function mnemonic_generate() {}
function key_derive() {}

module.exports = { key_derive, mnemonic_generate }
