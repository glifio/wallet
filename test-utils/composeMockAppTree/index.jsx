/* eslint-disable react/prop-types */
import React, { useReducer } from 'react'
import { Provider } from 'react-redux'
import { Converter } from '@openworklabs/filecoin-number'

import { NetworkCheck } from '../../lib/check-network'
import BalancePoller from '../../lib/update-balance'
import { ConverterContext } from '../../lib/Converter'
import { theme, ThemeProvider } from '../../components/Shared'
import { initializeStore } from '..'
import walletProviderReducer, {
  initialState as walletProviderInitialState
} from '../../WalletProvider/state'
import { WasmContext } from '../../lib/WasmLoader'
import { WalletProviderContext } from '../../WalletProvider'
import { mockWalletProviderInstance } from '../mocks/mock-wallet-provider'
import createMockWalletProviderContextFuncs from './createWalletProviderContextFuncs.js'
import mockWalletSubproviders from '../mocks/mock-wallet-subproviders'
import { presets, composeWalletProviderState } from './composeState'

/**
 * VALID OPTIONS:
 * mockConverterInstance (an object that stubs the converter functionality)
 * walletProviderReducer (a reducer to stub the state handling of the wallet provider)
 * walletProviderInitialState(a reducer to stub the state of the walletProvider context)
 * walletProviderDispatch(a dispatcher to mock/read calls to the walletProvider's state)
 * pathname
 * query (usually { network: 'f' } or { network: 't' })
 */

export default (statePreset = 'preOnboard', options = {}) => {
  const state = options.state || presets[statePreset]
  const store = initializeStore(state)

  const pathname = options.pathname || '/wallet'
  const query = options.query || { network: 't' }

  const Tree = ({ children }) => {
    const [initialWalletProviderState, walletProviderDispatch] = useReducer(
      options.reducer || walletProviderReducer,
      options.walletProviderInitialState || walletProviderInitialState
    )

    const walletProviderState = composeWalletProviderState(
      initialWalletProviderState,
      statePreset
    )

    const mockWalletProviderContextFuncs = createMockWalletProviderContextFuncs(
      options.walletProviderDispatch || walletProviderDispatch
    )

    return (
      <Provider store={store}>
        <WasmContext.Provider
          value={{
            generateMnemonic: jest
              .fn()
              .mockImplementation(
                () =>
                  'slender spread awkward chicken noise useful thank dentist tip bronze ritual explain version spot collect whisper glow peanut bus local country album punch frown'
              )
          }}
        >
          <ConverterContext.Provider
            value={{
              converter: new Converter(),
              converterError: options.converterError || null
            }}
          >
            <WalletProviderContext.Provider
              value={{
                state: walletProviderState,
                dispatch:
                  options.walletProviderDispatch || walletProviderDispatch,
                walletSubproviders: {
                  ...mockWalletSubproviders
                },
                ...mockWalletProviderContextFuncs
              }}
            >
              <NetworkCheck
                networkFromRdx={store.getState().network}
                pathname={pathname}
                query={query}
                switchNetwork={jest.fn()}
              />
              <BalancePoller />
              <ThemeProvider theme={theme}>{children}</ThemeProvider>
            </WalletProviderContext.Provider>
          </ConverterContext.Provider>
        </WasmContext.Provider>
      </Provider>
    )
  }

  return {
    Tree,
    store,
    walletProvider: mockWalletProviderInstance
  }
}
