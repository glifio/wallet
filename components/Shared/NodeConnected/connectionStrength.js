/**
 * This function basically computes a "connection score": 0 - disconnected, 1 bad node, 2 good to go
 *
 * A "connection" is a general health check - it consists of:
 * - do we have HTTP access to this node URL (is the node online?)
 * - if connected, is the node synced to the canonical chain?
 */

import axios from 'axios'
import LotusRPCEngine from '@openworklabs/lotus-jsonrpc-engine'

import { FILSCOUT, FILSCAN } from '../../../constants'
import reportError from '../../../utils/reportError'

const TIPSET_PADDING = 15

const filscoutHeight = async () => {
  try {
    const { data } = await axios.get(`${FILSCOUT}/network/overview`)
    if (data.code !== 200) {
      return 0
    }

    return Number(data.data.statistic.tipset_height)
  } catch (err) {
    return 0
  }
}

const filscanHeight = async () => {
  try {
    const { data } = await axios.get(`${FILSCAN}/BaseInformation`)
    if (data.res.code !== 3) {
      return 0
    }
    return Number(data.data.tipset_height)
  } catch (err) {
    return 0
  }
}

const getHeightFromRPC = async (apiAddress, token) => {
  try {
    const client = new LotusRPCEngine({
      apiAddress,
      token
    })
    const chainHead = await client.request('ChainHead')
    return Number(chainHead.Height)
  } catch (err) {
    return 0
  }
}

const filterOutBadNodes = heights => heights.filter(height => height > 0)

const collectHeights = async () => {
  const calls = []

  for (let i = 1; i <= 5; i += 1) {
    calls.push(getHeightFromRPC(`https://node.glif.io/0${i}/rpc/v0`))
  }

  calls.push(filscanHeight(), filscoutHeight())

  const heights = await Promise.all(calls)
  return filterOutBadNodes(heights)
}

// simple compute - sort the array,
// see if the node is TIPSET_PADDING blocks + than the highest height, or
// TIPSET_PADDING blocks - the lowest height
const computeScore = (height, heightsToCompare) => {
  heightsToCompare.sort((a, b) => a - b)
  const nodeFarAhead =
    height > heightsToCompare[heightsToCompare.length - 1] + TIPSET_PADDING
  const nodeFarBehind = height < heightsToCompare[0] - TIPSET_PADDING
  return nodeFarAhead || nodeFarBehind ? 1 : 2
}

export default async (apiAddress, token = '') => {
  const client = new LotusRPCEngine({
    apiAddress,
    token
  })

  const heights = await collectHeights()
  // if we cant make any comparisons, were assuming the network is unhealthy
  if (heights.length < 2) return 1

  try {
    const chainHead = await client.request('ChainHead')
    return computeScore(chainHead.Height, heights)
  } catch (err) {
    reportError(
      'components/Shared/NodeConnected/connectionStrength.js:1',
      false,
      err.message,
      err.stack
    )
    return 0
  }
}
