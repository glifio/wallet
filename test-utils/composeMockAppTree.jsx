/* eslint-disable react/prop-types */
import React, { useReducer } from 'react'
import cloneDeep from 'lodash.clonedeep'

import WalletProvider from '@openworklabs/filecoin-wallet-provider'
import { FilecoinNumber } from '@openworklabs/filecoin-number'
import { Provider } from 'react-redux'

import { NetworkCheck } from '../lib/check-network'
import BalancePoller from '../lib/update-balance'
import { ConverterContext } from '../lib/Converter'
import { theme, ThemeProvider } from '../components/Shared'
import { initializeStore } from './index'
import { initialState } from '../store/states'
import walletProviderReducer, {
  initialState as walletProviderInitialState,
  setWalletType,
  setError,
  resetLedgerState,
  resetState
} from '../WalletProvider/state'
import { WalletProviderContext } from '../WalletProvider'
import createPath from '../utils/createPath'
import { IMPORT_MNEMONIC } from '../constants'

const useRouter = jest.spyOn(require('next/router'), 'useRouter')

const mockRouter = jest.fn(() => {})
useRouter.mockImplementation(() => ({
  query: 'network=t',
  push: mockRouter,
  replace: mockRouter
}))

jest.mock('@openworklabs/filecoin-wallet-provider')
const mockGetAccounts = jest
  .fn()
  .mockImplementation(() =>
    Promise.resolve(['t1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq'])
  )

const mockGetBalance = jest
  .fn()
  .mockImplementation(() => Promise.resolve(new FilecoinNumber('1', 'fil')))

WalletProvider.mockImplementation(() => {
  return {
    wallet: {
      getAccounts: mockGetAccounts,
      sign: jest.fn().mockImplementation(() => 'xxxyyyyzzzz')
    },
    getBalance: mockGetBalance,
    getNonce: jest.fn().mockImplementation(() => 0),
    estimateGas: jest
      .fn()
      .mockImplementation(() => new FilecoinNumber('122', 'attofil')),
    sendMessage: jest.fn().mockImplementation(() => 'QmZCid!')
  }
})

const mockWalletProviderInstance = new WalletProvider()

const presets = {
  preOnboard: cloneDeep(initialState),
  postOnboard: cloneDeep({
    ...initialState,
    wallets: [
      {
        address: 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi',
        balance: new FilecoinNumber('1', 'fil'),
        path: createPath(1, 0)
      }
    ],
    selectedWalletIdx: 0
  })
}

const composeWalletProviderState = (initialWalletProviderState, preset) => {
  switch (preset) {
    case 'postOnboard': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance
      })
    }
    default:
      return initialWalletProviderState
  }
}

/**
 * VALID OPTIONS:
 * mockConverterInstance (an object that stubs the converter functionality)
 * walletProviderReducer (a reducer to stub the state handling of the wallet provider)
 * walletProviderInitialState(a reducer to stub the state of the walletProvider context)
 * pathname
 * query (usually { network: 'f' } or { network: 't' })
 */

export default (statePreset = 'preOnboard', options = {}) => {
  const state = options.state || presets[statePreset]
  const store = initializeStore(state)

  const mockConverterInstance = options.mockConverterInstance || {
    cacheConversionRate: jest.fn(),
    toFIL: jest.fn().mockImplementation(amount => {
      return new FilecoinNumber(amount, 'fil').dividedBy(5)
    }),
    fromFIL: jest.fn().mockImplementation(amount => {
      return new FilecoinNumber(amount, 'fil').multipliedBy(5)
    })
  }

  const pathname = options.pathname || '/wallet'
  const query = options.query || { network: 't' }

  const Tree = ({ children }) => {
    const [initialWalletProviderState, walletProviderDispatch] = useReducer(
      options.reducer || walletProviderReducer,
      options.walletProviderInitialState || walletProviderInitialState
    )

    const walletProviderState = composeWalletProviderState(
      initialWalletProviderState,
      statePreset
    )
    return (
      <Provider store={store}>
        <ConverterContext.Provider
          value={{
            converter: mockConverterInstance,
            converterError: options.converterError || null
          }}
        >
          <WalletProviderContext.Provider
            value={{
              state: walletProviderState,
              dispatch: walletProviderDispatch,
              fetchDefaultWallet: jest.fn().mockImplementation(),
              setWalletError: jest
                .fn()
                .mockImplementation(errorMessage =>
                  walletProviderDispatch(setError(errorMessage))
                ),
              setWalletType: jest
                .fn()
                .mockImplementation(walletType =>
                  walletProviderDispatch(setWalletType(walletType))
                ),
              setLedgerProvider: jest.fn(),
              connectLedger: jest.fn(),
              resetLedgerState: jest.fn().mockImplementation(() => {
                walletProviderDispatch(resetLedgerState())
              }),
              resetState: jest.fn().mockImplementation(() => {
                walletProviderDispatch(resetState())
              })
            }}
          >
            <NetworkCheck
              networkFromRdx={store.getState().network}
              pathname={pathname}
              query={query}
              switchNetwork={jest.fn()}
            />
            <BalancePoller />
            <ThemeProvider theme={theme}>{children}</ThemeProvider>
          </WalletProviderContext.Provider>
        </ConverterContext.Provider>
      </Provider>
    )
  }

  return { Tree, store, walletProvider: mockWalletProviderInstance }
}
