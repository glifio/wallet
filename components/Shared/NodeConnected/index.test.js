jest.mock('axios')
jest.mock('@openworklabs/lotus-jsonrpc-engine')

import { cleanup, render, act, wait } from '@testing-library/react'
import axios from 'axios'
import LotusRPCEngine from '@openworklabs/lotus-jsonrpc-engine'
import connectionStrength from './connectionStrength'
import NodeConnectedWidget from '.'
import ThemeProvider from '../ThemeProvider'
import { flushPromises } from '../../../test-utils'

describe('nodeConnectedWidget', () => {
  describe('nodeConnectedWidget UI Component', () => {
    afterEach(cleanup)
    beforeEach(() => {
      jest.useFakeTimers()
      jest.clearAllMocks()
      LotusRPCEngine.mockImplementation(() => {
        return {
          request: jest.fn(() => ({ Height: 1000 }))
        }
      })
      // filscan gets called first, so we mock its return val first
      axios.get.mockResolvedValueOnce({
        data: {
          res: {
            code: 3
          },
          data: { tipset_height: 1000 }
        }
      })

      // then filscout
      axios.get.mockResolvedValueOnce({
        data: {
          code: 200,
          data: {
            statistic: {
              tipset_height: 1000
            }
          }
        }
      })
    })

    test('it renders connecting state', () => {
      const res = render(
        <ThemeProvider>
          <NodeConnectedWidget apiAddress='https://node.glif.io/02/rpc/v0' />
        </ThemeProvider>
      )
      expect(res.container.firstChild).toMatchSnapshot()
    })

    test('it renders connected state', async () => {
      const onChangeSpy = jest.fn()
      let res
      await act(async () => {
        res = render(
          <ThemeProvider>
            <NodeConnectedWidget
              onConnectionStrengthChange={onChangeSpy}
              apiAddress='https://node.glif.io/02/rpc/v0'
            />
          </ThemeProvider>
        )
        await flushPromises()
        jest.runAllTimers()
      })
      expect(res.container.firstChild).toMatchSnapshot()
      expect(onChangeSpy).toHaveBeenCalledWith(2)
    })

    test('it pushes to the error page when node disconnects', async () => {
      LotusRPCEngine.mockImplementation(() => {
        return { request: jest.fn(() => ({ Height: 1000 })) }
      }).mockImplementationOnce(() => {
        return { request: jest.fn(() => ({ Height: 100 })) }
      })
      // filscan gets called first, so we mock its return val first
      axios.get.mockResolvedValueOnce({
        data: {
          res: {
            code: 3
          },
          data: { tipset_height: 1000 }
        }
      })

      // then filscout
      axios.get.mockResolvedValueOnce({
        data: {
          code: 200,
          data: {
            statistic: {
              tipset_height: 1000
            }
          }
        }
      })

      const onChangeSpy = jest.fn()
      let res
      await act(async () => {
        res = render(
          <ThemeProvider>
            <NodeConnectedWidget
              onConnectionStrengthChange={onChangeSpy}
              apiAddress='https://node.glif.io/02/rpc/v0'
            />
          </ThemeProvider>
        )
        await flushPromises()
        jest.runAllTimers()
      })
      expect(res.container.firstChild).toMatchSnapshot()
      expect(onChangeSpy).toHaveBeenCalledWith(1)
    })
  })

  describe('connectionStrength', () => {
    beforeEach(jest.clearAllMocks)

    test('it returns 2 when the Chain is synced ', async () => {
      LotusRPCEngine.mockImplementation(() => {
        return {
          request: jest.fn(() => ({ Height: 1000 }))
        }
      })
      // filscan gets called first, so we mock its return val first
      axios.get.mockResolvedValueOnce({
        data: {
          res: {
            code: 3
          },
          data: { tipset_height: 1000 }
        }
      })

      // then filscout
      axios.get.mockResolvedValueOnce({
        data: {
          code: 200,
          data: {
            statistic: {
              tipset_height: 1000
            }
          }
        }
      })

      const nodeURL = 'https://proxy.openworklabs.com/rpc/v0'
      const strength = await connectionStrength(nodeURL)
      expect(strength).toBe(2)
    })

    test('it will return a score of 1 if the node is not connected to the right network', async () => {
      // request should stay the same for the first 5 nodes, then once for our actual call
      LotusRPCEngine.mockImplementation(() => {
        return { request: jest.fn(() => ({ Height: 1000 })) }
      }).mockImplementationOnce(() => {
        return { request: jest.fn(() => ({ Height: 100 })) }
      })
      // filscan gets called first, so we mock its return val first
      axios.get.mockResolvedValueOnce({
        data: {
          res: {
            code: 3
          },
          data: { tipset_height: 1000 }
        }
      })

      // then filscout
      axios.get.mockResolvedValueOnce({
        data: {
          code: 200,
          data: {
            statistic: {
              tipset_height: 1000
            }
          }
        }
      })

      const nodeURL = 'https://proxy.openworklabs.com/rpc/v0'
      const strength = await connectionStrength(nodeURL)
      expect(strength).toBe(1)
    })
  })
})
