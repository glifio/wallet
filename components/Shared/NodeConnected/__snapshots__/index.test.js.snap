// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodeConnectedWidget nodeConnectedWidget UI Component it pushes to the error page when node disconnects 1`] = `
.c0 {
  min-width: 0;
  box-sizing: border-box;
  border: 1px solid;
  border-color: #C4C4C4;
  border-radius: 8px;
  padding: 8px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  min-width: 160px;
  width: 160px;
  height: 48px;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.4;
  font-family: RT-Alias-Grotesk;
  text-align: center;
  display: inline;
  margin: 0;
}

.c1 {
  height: 8px;
  width: 8px;
  background-color: #FC6D6F;
  border-radius: 50%;
  display: inline-block;
  margin: 4px;
  -webkit-animation: iSdllx 0.24s linear;
  animation: iSdllx 0.24s linear;
}

<div
  class="c0"
  display="inline-flex"
  height="6"
  width="9"
>
  <span
    class="c1"
  />
  <p
    class="c2"
    display="inline"
    font-family="RT-Alias-Grotesk"
    font-size="3"
    font-weight="400"
  >
    Disconnected
  </p>
</div>
`;

exports[`nodeConnectedWidget nodeConnectedWidget UI Component it renders connected state 1`] = `
.c0 {
  min-width: 0;
  box-sizing: border-box;
  border: 1px solid;
  border-color: #C4C4C4;
  border-radius: 8px;
  padding: 8px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  min-width: 160px;
  width: 160px;
  height: 48px;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.4;
  font-family: RT-Alias-Grotesk;
  text-align: center;
  display: inline;
  margin: 0;
}

.c1 {
  height: 8px;
  width: 8px;
  background-color: #1AD08F;
  border-radius: 50%;
  display: inline-block;
  margin: 4px;
  -webkit-animation: iSdllx 0.24s linear;
  animation: iSdllx 0.24s linear;
}

<div
  class="c0"
  display="inline-flex"
  height="6"
  width="9"
>
  <span
    class="c1"
  />
  <p
    class="c2"
    display="inline"
    font-family="RT-Alias-Grotesk"
    font-size="3"
    font-weight="400"
  >
    Connected
  </p>
</div>
`;

exports[`nodeConnectedWidget nodeConnectedWidget UI Component it renders connecting state 1`] = `
.c0 {
  min-width: 0;
  box-sizing: border-box;
  border: 1px solid;
  border-color: #C4C4C4;
  border-radius: 8px;
  padding: 8px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  min-width: 160px;
  width: 160px;
  height: 48px;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  font-size: 1.25rem;
  font-weight: 400;
  line-height: 1.4;
  font-family: RT-Alias-Grotesk;
  text-align: center;
  display: inline;
  margin: 0;
}

.c1 {
  height: 8px;
  width: 8px;
  background-color: #FCA703;
  border-radius: 50%;
  display: inline-block;
  margin: 4px;
  -webkit-animation: iSdllx 0.24s linear;
  animation: iSdllx 0.24s linear;
}

<div
  class="c0"
  display="inline-flex"
  height="6"
  width="9"
>
  <span
    class="c1"
  />
  <p
    class="c2"
    display="inline"
    font-family="RT-Alias-Grotesk"
    font-size="3"
    font-weight="400"
  >
    Connecting..
  </p>
</div>
`;
